source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -30,10 +33,129 @@ namespace Tls {
+          ENVOY_LOG(warn, "Duplicate domain '{}' in SPIFFE bundle map", domain_name);
+        } else {
+          spiffeDataPtr->trust_bundle_stores[domain_name] = X509StorePtr(X509_STORE_new());
+        }
+
+        ENVOY_LOG(info, "Loading domain '{}' from SPIFFE bundle map", domain_name);
+
+        const auto keys = domain_object.getObjectArray("keys");
+
+        if (keys.empty()) {
+          ENVOY_LOG(error, "No keys found in SPIFFE bundle for domain '{}'", domain_name);
+          error = true;
+          return false;
+        }
+
+        ENVOY_LOG(info, "Found '{}' keys for domain '{}'", keys.size(), domain_name);
+
+        for (const auto& key : keys) {
+          if (key->getString("use") == "x509-svid") {
+            const auto& certs = key->getStringArray("x5c");
+            for (const auto& cert : certs) {
+              std::string decoded_cert = Envoy::Base64::decode(cert);
+              if (decoded_cert.empty()) {
+                ENVOY_LOG(error, "Empty cert decoded in domain '{}'", domain_name);
+                error = true;
+                return false;
+              }
+
+              const unsigned char* cert_data =
+                  reinterpret_cast<const unsigned char*>(decoded_cert.data());
+              bssl::UniquePtr<X509> x509(d2i_X509(nullptr, &cert_data, decoded_cert.size()));
+              if (!x509) {
+                ENVOY_LOG(error, "Failed to create x509 object while loading certs in domain '{}'",
+                          domain_name);
+                error = true;
+                return false;
+              }
+              if (X509_STORE_add_cert(spiffeDataPtr->trust_bundle_stores[domain_name].get(),
+                                      x509.get()) != 1) {
+                ENVOY_LOG(error, "Failed to add x509 object while loading certs for domain '{}'",
+                          domain_name);
+                error = true;
+                return false;
+              }
+              spiffeDataPtr->ca_certs.push_back(std::move(x509));
+            }
+          }
+        }
+
+        return true;
+      });
+
+  if (error) {
+    return nullptr;
+  }
+
+  ENVOY_LOG(info, "Successfully loaded SPIFFE bundle map");
+  return spiffeDataPtr;
+}
+
+void SPIFFEValidator::initializeCertificateRefresh(
+    Server::Configuration::CommonFactoryContext& context) {
+  file_watcher_ = context.mainThreadDispatcher().createFilesystemWatcher();
+  THROW_IF_NOT_OK(file_watcher_->addWatch(
+      trust_bundle_file_name_, Filesystem::Watcher::Events::Modified, [this](uint32_t) {
+        ENVOY_LOG(info, "Updating SPIFFE bundle map from file '{}'", trust_bundle_file_name_);
+        std::ifstream file(trust_bundle_file_name_);
+        if (file.fail()) {
+          ENVOY_LOG(error, "Failed to open SPIFFE bundle map file '{}'", trust_bundle_file_name_);
+          return absl::OkStatus();
+        }
+
+        std::stringstream file_string;
+        file_string << file.rdbuf();
```

Comment:
Please use `fileSystem()` from the `Api::Api` to read file rather than use the `ifstream` directly.